jax_distribution/systems/rec_mappo_rware.py

```diff
@@ -0,0 +1,737 @@
+
+        masked_logits = jnp.where(
+            observation.action_mask,
+            actor_output,
+            jnp.finfo(jnp.float32).min,
+        )
+
+        pi = distrax.Categorical(logits=masked_logits)
+
+        critic_embedding = nn.Dense(
+            128, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0)
+        )(observation.global_state)
+        critic_embedding = nn.relu(critic_embedding)
+
+        critic_rnn_in = (critic_embedding, done)
+        critic_hidden_state, critic_embedding = ScannedRNN()(
+            critic_hidden_state, critic_rnn_in
+        )
+
+        critic = nn.Dense(128, kernel_init=orthogonal(2), bias_init=constant(0.0))(
+            critic_embedding
+        )
+        critic = nn.relu(critic)
+        critic = nn.Dense(1, kernel_init=orthogonal(1.0), bias_init=constant(0.0))(
+            critic
+        )
+
+        hiddens = (policy_hidden_state, critic_hidden_state)
+
+        return hiddens, pi, jnp.squeeze(critic, axis=-1)
+
+
+def get_learner_fn(
+    env: jumanji.Environment, apply_fn: Callable, update_fn: Callable, config: Dict
+) -> Callable:
+    """Get the learner function."""
+
+    def _update_step(
+        learner_state: RNNLearnerState, _: Any
+    ) -> Tuple[RNNLearnerState, Tuple]:
+        """A single update of the network.
+
+        This function steps the environment and records the trajectory batch for
+        training. It then calculates advantages and targets based on the recorded
+        trajectory and updates the actor and critic networks based on the calculated
+        losses.
+
+        Args:
+            learner_state (NamedTuple):
+                - params (FrozenDict): The current model parameters.
+                - opt_state (OptState): The current optimizer state.
+                - rng (PRNGKey): The random number generator state.
+                - env_state (State): The environment state.
+                - last_timestep (TimeStep): The last timestep in the current trajectory.
+                - last_done (bool): Whether the last timestep was a terminal state.
+                - hstates (chex.Array): The hidden state of the RNN.
+            _ (Any): The current metrics info.
+        """
+
+        def _env_step(
+            learner_state: RNNLearnerState, _: Any
+        ) -> Tuple[RNNLearnerState, PPOTransition]:
+            """Step the environment."""
+            (
+                params,
+                opt_state,
+                rng,
+                env_state,
+                last_timestep,
+                last_done,
+                hstates,
+            ) = learner_state
+
+            rng, policy_rng = jax.random.split(rng)
+
+            # Add a batch dimension to the observation.
+            batched_observation = jax.tree_util.tree_map(
+                lambda x: x[np.newaxis, :], last_timestep.observation
+            )
+            ac_in = (
+                batched_observation,
+                last_done[:, 0][np.newaxis, :],
```

Comment:
Only seeing this now, but I think in other places we use `None` to add a new axis, I actually quite like `np.newaxis` as it's clearer what is happening, but we should just stay consistent, so either change these to `None` or change the others to `np.newaxis`

Reply:
Having gone through the PR I don't see adding a new axis with `None` anywhere, so happy for you to keep things as they are. Just have a look in the other files as I feel like I saw it somewhere :shrug:

Reply:
We changed them to jnp.newaxis